[docs] Add mockgen reference to docs (#26978)
expo/expo: docs/pages/modules/mocking.mdx

The recommended way to write unit tests for an Expo project is to use [Jest](https://jestjs.io/) and the `jest-expo` preset.

To write a unit test for an app that uses native code, you need to mock native calls.
The term **Mocking** means to replace the actual implementation of a function with a fake version that does not perform any actions. This approach is useful for running unit tests on a local computer as it involves bypassing the need for native code which is only available on a device, and any code that calls native functions on a local machine will not work.

Expo SDK includes a set of default mocks for each of our community packages. You can also mock any JS code yourself using built-in Jest APIs such as [mock functions](https://jestjs.io/docs/mock-functions).

However, to provide default mocks in your Expo Module, we offer a method to bundle them. This ensures that when your module user runs unit tests, they will automatically use a mocked implementation.

## Providing mocks for a module

Create a file with the same name as the native module you want to mock and place it in your module's **mocks** directory. Make sure to export the mock implementation from this file.

> The `jest-expo` preset will automatically return the exported functions as a result of a `requireNativeModule` call when running during a unit test.

Reviewer: ```suggestion
The `jest-expo` preset will automatically return the exported functions because of a `requireNativeModule` call when running during a unit test.
```
Suggestion to fix vale warning:

![CleanShot 2024-02-12 at 15 06 47@2x](https://github.com/expo/expo/assets/10234615/dad12ec0-270f-40c7-b505-71ffe85f8c76)

---
title: 'Mocking native calls in Expo modules'
sidebar_title: Mocking
description: Learn about mocking native calls in Expo modules.
---

import { BoxLink } from '~/ui/components/BoxLink';
import { Grid01Icon } from '@expo/styleguide-icons';
import { Terminal } from '~/ui/components/Snippet';

The recommended way to write unit tests for an Expo project is to use [Jest](https://jestjs.io/) and the `jest-expo` preset.

<BoxLink
  title="Unit testing"
  description="Learn how to set up and configure the jest-expo package to write unit tests and snapshot tests for a project."
  href="/develop/unit-testing/"
  Icon={Grid01Icon}

> />

Reviewer: Let's move this below tip and add a new section for it. As a reader, I'm a bit confused by seeing this as the first thing on the page and after the boxlink I see context on writing unit test for native code.

Author: That works, I was thinking something like: here's how to mock in general, and here's a specific case, but your suggestion also makes sense.

Reviewer: Thanks! 

May be we can add a generic section like "More" and in the description of the box link mention that for a more general purpose mocking, see our unit testing guide.

To unit test an application that uses native code, you will need to mock native calls.
Mocking means to replace the actual implementation of a function with a fake implementation (that usually does nothing).
This is useful, because unit tests run only on your computer, and without the native code running on the mobile device any code calling native functions will not work.

The Expo SDK includes a set of default mocks for each of our community packages. You can also mock any JS code yourself using built-in Jest APIs such as [mock functions](https://jestjs.io/docs/mock-functions).

However, if you want to provide default mocks in your Expo Module, we provide a way to bundle these mocks with your module. This way, when a user runs unit tests, they will automatically use a mocked implementation.

# Providing mocks for your module

Create a file with the same name as the native module you want to mock, and place it in a `mocks` directory of your module. Export your mock implementation from this file.
The `jest-expo` preset will automatically return the exported functions as a result of a `requireNativeModule` call when running during a unit test.

> As an example, the `expo-clipboard` module has a native module called `ExpoClipboard`. To mock this module, you would create a file called `ExpoClipboard.ts` in a `mocks` directory in your module.

Reviewer: ```suggestion
For example, the `expo-clipboard` library has a native module called `ExpoClipboard`. You will create a **ExpoClipboard.ts** in the **mocks** directory to mock it.
```

- Let's be concistent here since we represent file/directory names as bold so we don't have to implicitly declare file names
- Since we've already told that file needs to be created in mocks directory, we don't have to repeat ourselves here.

> ```tsx

Reviewer: ````suggestion
```ts ExpoClipboard.ts
````

export async function hasStringAsync(): Promise<boolean> {
  return false;
}
```

> Now, in a unit test, calling `ExpoClipboard.hasStringAsync()` will return `false`.

Reviewer: ```suggestion
Now, in a unit test, calling `ExpoClipboard.hasStringAsync()` returns `false`.
```

> # Automatic generation of mocks

Reviewer: ```suggestion
## Automatic generation of mocks
```

> Maintaining mocks for native modules can be a lot of work, especially if the native module has many methods. To make this easier, we provide a script that will automatically generate mocks for all native functions in your module.

Reviewer: ```suggestion
Maintaining mocks for native modules can be a lot of work if the native module has multiple methods. To make this easier, we provide a script that will automatically generate mocks for all native functions in your module.
```

> It works for generating mocks in TypeScript and JavaScript, based on the Swift implementation in your module.

Reviewer: ```suggestion
It works for generating mocks in TypeScript and JavaScript based on the Swift implementation in your module.
```

> To use this script you need to have [SourceKitten](https://github.com/jpsim/SourceKitten) framework installed and run the `generate-ts-mocks` command in your module.

Reviewer: ```suggestion
To use this script, you have to install [SourceKitten](https://github.com/jpsim/SourceKitten) framework and run the `generate-ts-mocks` command in your module.
```

<Terminal
  cmdCopy="brew install sourcekitten && npx expo-modules-test-core generate-ts-mocks"
  cmd={['$ brew install sourcekitten', '$ npx expo-modules-test-core generate-ts-mocks']}
/>

> This will generate a file called `ExpoModuleName.ts` in the `mocks` directory of your module. This file will contain a mock implementation for each native method and view in your module.

Reviewer: ```suggestion
The command above will generate **ExpoModuleName.ts** in the **mocks** directory of your module. It will contain a mock implementation for each native method and view in your module.
```
